### ISSUE TITLE
Errors when starting flannel: "Error adding route ... file exists"

### ISSUE THREAD

**Issue by DreadPirateShawn** [coreos/flannel#333]
Whenever we start flannel on multiple hosts, we see the following error logs:

Host: **kube-minion1**
```
/opt/flannel/bin/flanneld --v=2 --ip-masq=true --subnet-file=/var/run/flannel/subnet.env --iface=eth0 --etcd-endpoints=http://localhost:4001 --etcd-prefix=/flannel
```

```
$ cat /var/run/flannel/subnet.env
FLANNEL_SUBNET=10.244.38.1/24
FLANNEL_MTU=1500
FLANNEL_IPMASQ=true
```

```
$ cat flanneld.ERROR 
Log file created at: 2015/09/29 20:47:29
Running on machine: kube-minion1
Binary: Built with gc go1.4.1 for linux/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
E0929 20:47:29.345384 20145 hostgw.go:140] Error adding route to 10.244.15.0/24 via 10.0.0.7: file exists
E0929 20:47:29.349848 20145 hostgw.go:140] Error adding route to 10.244.19.0/24 via 10.0.0.4: file exists
E0929 20:47:29.349966 20145 hostgw.go:140] Error adding route to 10.244.84.0/24 via 10.0.0.2: file exists
```

Host: **kube-minion2**
```
/opt/flannel/bin/flanneld --ip-masq=true --subnet-file=/var/run/flannel/subnet.env --iface=eth0 --etcd-endpoints=http://localhost:4001 --etcd-prefix=/flannel
```

```
$ cat /var/run/flannel/subnet.env 
FLANNEL_SUBNET=10.244.19.1/24
FLANNEL_MTU=1500
FLANNEL_IPMASQ=true
```

```
$ cat flanneld.ERROR 
Log file created at: 2015/09/29 20:47:27
Running on machine: kube-minion2
Binary: Built with gc go1.4.1 for linux/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
E0929 20:47:27.615004 17655 hostgw.go:140] Error adding route to 10.244.15.0/24 via 10.0.0.7: file exists
E0929 20:47:27.616871 17655 hostgw.go:140] Error adding route to 10.244.38.0/24 via 10.0.0.3: file exists
E0929 20:47:27.616998 17655 hostgw.go:140] Error adding route to 10.244.84.0/24 via 10.0.0.2: file exists
E0929 20:47:28.878947 17655 hostgw.go:140] Error adding route to 10.244.38.0/24 via 10.0.0.3: file exists
```

Host: **kube-minion3**

```
/opt/flannel/bin/flanneld --ip-masq=true --subnet-file=/var/run/flannel/subnet.env --iface=eth0 --etcd-endpoints=http://localhost:4001 --etcd-prefix=/flannel
```

```
$ cat /var/run/flannel/subnet.env
FLANNEL_SUBNET=10.244.15.1/24
FLANNEL_MTU=1500
FLANNEL_IPMASQ=true
```

```
$ cat flanneld.ERROR 
Log file created at: 2015/09/29 20:41:32
Running on machine: kube-minion3
Binary: Built with gc go1.4.1 for linux/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
E0929 20:41:32.019905 01984 hostgw.go:140] Error adding route to 10.244.19.0/24 via 10.0.0.4: file exists
E0929 20:41:32.031599 01984 hostgw.go:140] Error adding route to 10.244.38.0/24 via 10.0.0.3: file exists
E0929 20:41:32.031685 01984 hostgw.go:140] Error adding route to 10.244.84.0/24 via 10.0.0.2: file exists
E0929 20:41:32.098188 01984 hostgw.go:140] Error adding route to 10.244.19.0/24 via 10.0.0.4: file exists
E0929 20:41:33.399350 01984 hostgw.go:140] Error adding route to 10.244.38.0/24 via 10.0.0.3: file exists
```

Functionality seems to work just fine, but the error logs seem concerning. Are there flag modifications that we should make to address this? or something that should be fixed in flannel going forward?

**Comment by eyakubovich** [comment]
These are just left from previous runs of flannel. flannel currently doesn't clean up on exit which is good and bad. It good because you can take it down and restart without sacrificing existing connectivity (except for udp backend). It's bad because we need to be careful to check for such errors and not report them. We'll address this but it should be benign.

**Comment by DreadPirateShawn** [comment]
Thanks for the clarification! The restart-without-losing-connectivity scenario makes sense.

Out of curiosity, apart from the default behavior which doesn't clean up, is there a way to tell flannel that you do want to clean up? That is, do the routes created by flannel remain in existence "forever", or is there a way to tell flannel to stop and remove the routes, to return the system to its original pre-flannel state?

**Comment by eyakubovich** [comment]
Not yet. But since we're doing more work on multi-network support and them being dynamic, it is something that we should add.

**Comment by philips** [comment]
@eyakubovich We would have to track this state outside of flannel on disk somewhere, right? There is no way to tag routes with metadata like "this was added by flannel".

**Comment by eyakubovich** [comment]
@philips I don't think there's a way to add a comment to a route like you can with the iptables rule. But if you clean up on exit, you still know the flannel network so you can match on the destination address. Obviously it's not robust in the face of segfaults and such.

For some backends, it cleans up old cruft on start.

**Comment by fnordahl** [comment]
I would suggest that tearing down the configuration on startup or exit could be harmfull. If a daemon can be restarted for updates or other service without affecting the data plane it will make it much easier for operators to service and it will also make the system seem much more robust and stable.

I have proposed a patch that will look up the routes before attempting to create them. Any existing equal routes will be left untouched and will not genereate errors. Existing routes with different Gw will be replaced.

What do you think @eyakubovich (Cc @philips )

**Comment by steveeJ** [comment]
@fnordahl 

> Any existing equal routes will be left untouched and will not genereate errors. Existing routes with different Gw will be replaced.

It seems like a step in the right direction.

Long-term I think it's worth the effort to have a synchronized state between the local kernel routes and flannel's global knowledge about the routes. If flannel is (re)started, it's a reasonable assumption can be made that all routes that match the flannel subnet have and will be managed by flannel.

**Comment by steveeJ** [comment]
Fixed by #458.

**Comment by fnordahl** [comment]
Thanks @steveeJ
